### ISSUE TITLE
Missed DeprecationWarning, PendingDeprecationWarning in NumPy tests

### ISSUE THREAD

**Issue by charris** [numpy/numpy#10185]
These show up release tests, and should raise errors in development, but they don't. We must be messing with the warnings queue somewhere.
```
test_nat_items (numpy.testing.tests.test_utils.TestEqual) ... /home/charris/Workspace/numpy.git/build/testenv/lib64/python2.7/site-packages/numpy/testing/nose_tools/utils.py:407: DeprecationWarning: elementwise == comparison failed; this will raise an error in the future.
  if not (desired == actual):
```
and
```
numpy.core.tests.test_multiarray.TestFormat.test_1d_format ... /home/charris/Workspace/numpy.git/build/testenv/lib64/python2.7/site-packages/numpy/core/tests/test_multiarray.py:7092: PendingDeprecationWarning: object.__format__ with a non-empty format string is deprecated
  return f(*args, **kwargs), None
```

**Comment by njsmith** [comment]
I think the first one is being missed because we only convert warnings into
errors if they occur in numpy code, and here it's nose that's actually
making the offending call. Not sure if there's anything to be done or not;
this case is a false negative, but if we start making all warnings into
errors then that will probably trigger some false positives sooner or later.

**Comment by seberg** [comment]
But through some error handling all is good when no warning is given (didn't quite track it down yet)? The second one, I don't see right now (probably it depends on py version or so).

**Comment by seberg** [comment]
Ah, the `_GenericTest` in `test_utils` catches AssertionErrors (which the DeprecationWarning when raised would be converted into probably. The strange thing is that the DeprecationWarning is converted to an AssertionError maybe, hmmm

**Comment by charris** [comment]
The first one should also be caught in the `assert_equal` function, as it involves `NaT == NaT`, which we should not actually reach.

**Comment by charris** [comment]
The second one is, I think, a Python `PendingDeprecationWarning` <looks>, yep, comes from
```
        # Could switch on python version here, but all we care about is
        # that the behaviour hasn't changed
        assert_equal(
            ret_and_exc(object.__format__, a, '30'),
            ret_and_exc('{:30}'.format, a)
        )
```
We should probably just remove that, as testing against `object.__format__` is going to be deprecated.

**Comment by seberg** [comment]
Well, I guess what causes this is the tests that compare `[NaT]` to `NaT`. @eric-wieser, Eric, you seemed to have put the replacement of FutureWarning and DeprecationWarning there at the end when fixing up my NaT mess, do you remember off the top of your head why that was necessary?

**Comment by seberg** [comment]
Probably can just remove that test, true, though I find it a bit odd that PendingDeprecationWarning did not show up before? The tests should error on pretty much all warnings? Or maybe nobody was testing against a new enough python since then?

**Comment by eric-wieser** [comment]
> as testing against object.__format__ is going to be deprecated.

Is that a python change?

**Comment by eric-wieser** [comment]
> Well, I guess what causes this is the tests that compare [NaT] to NaT.

I suppose this is in `assert_array_equal`, since one of the elements is an array?

> do you remember off the top of your head why that was necessary?

It was to allow `assert_equal` to be used on types with no comparison defined without warning

**Comment by charris** [comment]
@seberg In order to see them, you need to have `ISRELEASED=True` in `setup.py` .

**Comment by eric-wieser** [comment]
I guess I decided that I didn't need to suppress the warnings, only when they were exceptions - I don't think I had a strong rationale behind that

**Comment by charris** [comment]
@eric-wieser , it is actually in `assert_equal` and is because the case of two `NaT` with different types is not caught as an error. It's an easy fix.

**Comment by eric-wieser** [comment]
I think the warning suppression is still needed there to deal with `assert_equal(void, datetime)` or other types where no `==` loop can be found

**Comment by eric-wieser** [comment]
The `__format__` thing is #9883 - also my doing

**Comment by charris** [comment]
> Is that a python change?
Yes, the message isn't in NumPy.

**Comment by seberg** [comment]
@charris, `python3 runtests.py --raise-warnings=release` should work as well. But in develop mode it should show up as an error, why doesn't it? (Maybe it does not matter, I might try to understand it myself later)

@eric-wieser why do we need to convert it to AssertionError? The other tests then catches AssertionErrors.
Unfortunately we cannot suppress warnings in those testing functions, because warning suppression is not thread safe, so if there is some change necessary there, it might end up being quite a dance.

**Comment by eric-wieser** [comment]
If we don't convert it to `AssertionError`, then we never see the message that tells us what the values are, which makes `assertEqual` kind of useless.

**Comment by charris** [comment]
@seberg It doesn't show up because the test is constructed to succeed if the assertion fails. Kind of weird, and not designed to work with deprecations.

**Comment by charris** [comment]
@eric-wieser I think we should simply re-raise the error if `assert_equal` fails due to `PendingDeprecationWarning` or `DeprecationWarning` in development mode. The current problem arises because the test that fails is using `AssertionError` to check non-equality. In release mode both of those are merely warnings instead of failures. We should be able to append the relevant message someplace in there.

**Comment by eric-wieser** [comment]
> because the test that fails is using AssertionError to check non-equality.

**Comment by charris** [comment]
Maybe, I'm going to take a look ...

**Comment by eric-wieser** [comment]
Hmm, actually this seems to be a pretty common thing to do in the meta-tests that test the assertion helpers.
I think the right thing to do is suppress the warning

**Comment by charris** [comment]
I think it already does the correct thing in release mode, the equality check goes through and a warning is issued. It is development mode where the warnings are converted to AssertionError that is the problem.

**Comment by charris** [comment]
Not that the `NaT == NaT` is fixed by handling it in the NaT comparison section, the problem is that we didn't catch that error during development.

**Comment by charris** [comment]
@eric-wieser I put up some fixes at #10193, if you want to do more with `assert_equal` you should comment there.
